File: commons/javafx/src/main/java/com/cedarsoft/commons/javafx/FontBuilder.java
package com.cedarsoft.commons.javafx;

import javax.annotation.Nonnull;

import javafx.scene.text.Font;
import javafx.scene.text.FontPosture;
import javafx.scene.text.FontWeight;

/**
 * A builder for {@link Font}s.
 */
public class FontBuilder {
  @Nonnull
  private String family;
  @Nonnull
  private FontWeight weight;
  @Nonnull
  private FontPosture posture;

  private double size;

  public FontBuilder() {
    this(Font.getDefault());
  }

  public FontBuilder(@Nonnull Font font) {
    family = font.getFamily();
    size = font.getSize();
    weight = FontWeight.NORMAL;
    posture = FontPosture.REGULAR;
  }

  @Nonnull
  public static FontBuilder create() {
    return new FontBuilder();
  }

  @Nonnull
  public static FontBuilder createWithFont(@Nonnull Font font) {
    return new FontBuilder(font);
  }

  @Nonnull
  public FontBuilder withFamily(@Nonnull String family) {
    this.family = family;
    return this;
  }

  @Nonnull
  public FontBuilder withWeight(@Nonnull FontWeight weight) {
    this.weight = weight;
    return this;
  }

  @Nonnull
  public FontBuilder withPosture(@Nonnull FontPosture posture) {
    this.posture = posture;
    return this;
  }

  @Nonnull
  public FontBuilder withSize(double size) {
    this.size = size;
    return this;
  }

  @Nonnull
  public Font build() {
    return Font.font(family, weight, posture, size);
  }
}

File: commons/javafx/src/main/kotlin/it/neckar/open/javafx/FontBuilder.kt
package it.neckar.open.javafx

import javafx.scene.text.Font
import javafx.scene.text.FontPosture
import javafx.scene.text.FontWeight
import javax.annotation.Nonnull

/**
 * A builder for [Font]s.
 */
class FontBuilder @JvmOverloads constructor(font: Font = Font.getDefault()) {
  private var family: String
  private var weight: FontWeight
  private var posture: FontPosture
  private var size: Double

  init {
    family = font.family
    size = font.size
    weight = FontWeight.NORMAL
    posture = FontPosture.REGULAR
  }

  fun withFamily(@Nonnull family: String): FontBuilder {
    this.family = family
    return this
  }

  fun withWeight(@Nonnull weight: FontWeight): FontBuilder {
    this.weight = weight
    return this
  }

  fun withPosture(@Nonnull posture: FontPosture): FontBuilder {
    this.posture = posture
    return this
  }

  fun withSize(size: Double): FontBuilder {
    this.size = size
    return this
  }

  fun build(): Font {
    return Font.font(family, weight, posture, size)
  }

  companion object {
    @JvmStatic
    fun create(): FontBuilder {
      return FontBuilder()
    }

    @JvmStatic
    fun createWithFont(@Nonnull font: Font): FontBuilder {
      return FontBuilder(font)
    }
  }
}
